Document Snowflake dynamic table scheduler config parameter (#8934)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/snowflake-configs.md

- **Time-based** &mdash; Provide a value of the form `<int> { seconds | minutes | hours | days }`. For example, if the dynamic table needs to be updated every 30 minutes, use `target_lag='30 minutes'`.
- **Downstream** &mdash; Applicable when the dynamic table is referenced by other dynamic tables. In this scenario, `target_lag='downstream'` allows for refreshes to be controlled at the target, instead of at each layer.

> <VersionBlock firstVersion="1.12">

Reviewer: @igorbelianski-cyber Since the requirement to set `scheduler: ENABLE` is versioned to 1.12+, does that mean it's a new requirement for `target_lag` to work? Are users able to use `target_lag` in 1.11 and below without explicitly enabling `scheduler`?

Author: no new requirements. (if text reads like we are introducing a new requirement lets fix it) 
   users can continue with their current model (just specify target_lag ) ,as is, without changes (   scheduler: ENABLE will be an assumed default ) and the resulting behavior will not change.  
   
   The new mode is (currently not supported ) users can choose not define target lag at all ( this will assume SCHEDULER =DISABLE)    
   
   in either option SCHEDULER =...... setting is not required

Author: i tried to rewrite this to make it more clear

Reviewer: Thanks for clarifying @igorbelianski-cyber ! I pushed the following changes in this commit: https://github.com/dbt-labs/docs.getdbt.com/pull/8934/changes/bff703cde3a4559b878480c410de35b2159ac4c3
- Removed the 1.12 version blocks since adding `scheduler` is not a new requirement. (If `scheduler` is only available starting 1.12, I'll add these back)
- Added `scheduler` to tables that are versioned for 1.11 and below.
- Converted the `target_lag and scheduler interaction` to the following matrix so it's easier to read:
    
<img width="848" height="466" alt="Screenshot 2026-04-16 at 8 39 37 PM" src="https://github.com/user-attachments/assets/0a055cf8-e54b-4af9-8927-9dc4f891e661" />

Let me know what you think!

Reviewer: Confirmed that `scheduler` is available 1.12+, so I added back the version blocks and removed `scheduler` from 1.11 tables in this commit: https://github.com/dbt-labs/docs.getdbt.com/pull/8934/changes/3ba677769a4d718b95d9f9a8319e30d40c940f03

Author: great, thank you

> :::caution target_lag requires scheduler to be set to ENABLE

Reviewer: ```suggestion
:::caution target_lag requires scheduler to be enabled
```

> `target_lag` is only applicable when [`scheduler`](#scheduler) is effectively set to `ENABLE`. You cannot set `target_lag` when `scheduler='DISABLE'`.

Reviewer: ```suggestion
You can only use `target_lag` when [`scheduler`](#scheduler) is set to `ENABLE`. You cannot set `target_lag` when `scheduler: DISABLE`.
```

Author: this is a bit misleading "ou can only use `target_lag` when scheduler is not provided or "ENABLED" 

may be something like : 
":::caution target_lag requires scheduler not to be explicitly disabled
`target_lag` is only applicable when [`scheduler`](#scheduler) is skipped or set to `ENABLE`. You cannot set `target_lag` when `scheduler='DISABLE'`.

WDYT?

Author: rewritten to address the spirit of the suggestion

> In dbt, if neither `target_lag` nor `scheduler` is provided, dbt defaults to `scheduler='DISABLE'` and manages refresh itself. If you specify `target_lag` and omit `scheduler`, dbt implicitly sets `scheduler='ENABLE'`, which opts into Snowflake-managed scheduling.

Reviewer: Some suggested changes for consistency:
```suggestion
In dbt, if neither `target_lag` nor `scheduler` is specified, dbt defaults to `scheduler: DISABLE` and manages refreshes itself. If you specify `target_lag` without `scheduler`, dbt implicitly sets `scheduler: ENABLE`, which opts into Snowflake-managed scheduling.
```

> In dbt, if neither `target_lag` nor `scheduler` is provided, dbt defaults to `scheduler='DISABLE'` and manages refresh itself. If you specify `target_lag` and omit `scheduler`, dbt implicitly sets `scheduler='ENABLE'`, which opts into Snowflake-managed scheduling.

Reviewer: Also @igorbelianski-cyber, I also noticed what looks like a contradiction in this callout:

It currently states:
> You cannot set `target_lag` when `scheduler='DISABLE'`.

But later it says:
> If you specify `target_lag` and omit `scheduler`, dbt implicitly sets `scheduler='ENABLE'`.

These seem to conflict -- the first implies a hard restriction, while the second implies dbt silently auto-resolves it. Could you clarify which behavior is actually implemented?

1. Does dbt raise an error if `target_lag` is set without `scheduler: ENABLE`?
2. Or does dbt automatically set `scheduler: ENABLE` when `target_lag` is specified?

Author: 1. Does dbt raise an error if target_lag is set without scheduler: ENABLE?   NO 
2.does dbt automatically set scheduler: ENABLE when target_lag is specified?  YES

Author: PTAL i think i cleaned it up

:::

</VersionBlock>

Learn more about `target_lag` in Snowflake's [docs](https://docs.snowflake.com/en/user-guide/dynamic-tables-refresh#understanding-target-lag). Please note that Snowflake supports a target lag of 1 minute or longer.

<VersionBlock firstVersion="1.12">

### Scheduler

> Controls whether Snowflake's background scheduler automatically refreshes the dynamic table, or whether dbt manages refreshes directly. Snowflake allows two options:

Reviewer: ```suggestion
The `scheduler` parameter controls whether the dynamic table is refreshed by Snowflake's background scheduler or by an external orchestrator (for example, dbt). Snowflake accepts two options:
```

> - **ENABLE** &mdash; Snowflake's built-in scheduler automatically refreshes the dynamic table based on the defined `target_lag`. Refreshes cascade across the dependency graph to maintain snapshot consistency. `target_lag` is **required** in this mode.

Reviewer: ```suggestion
- **ENABLE** &mdash; Snowflake's built-in scheduler automatically refreshes the dynamic table based on the defined `target_lag`. Refreshes cascade across the dependency graph to maintain snapshot consistency. Setting `target_lag` is _required_ when using this option. 
```

> - **DISABLE** &mdash; The dynamic table is excluded from automatic Snowflake background refresh. Refreshing must be triggered manually or through orchestration external to Snowflake (for example, by a `dbt run` that executes `ALTER DYNAMIC TABLE ... REFRESH`). `target_lag` **cannot** be set in this mode.

Reviewer: ```suggestion
- **DISABLE** &mdash; The dynamic table is excluded from Snowflake's automatic background refresh. You must trigger refreshes manually or through orchestration external to Snowflake (for example, by a `dbt run` that executes `ALTER DYNAMIC TABLE ... REFRESH`). When using this option, specifying `target_lag` will result in an error. 
```

Author: that's not  quite right ", specifying `target_lag`" will result in Dynamic table background scheduler taking over ( and DBT not refreshing it itself) . Not a failure just different behavior

Author: fixed PTAL

:::info dbt default differs from Snowflake's native default

> In Snowflake's native DDL, omitting `SCHEDULER` defaults to scheduler-enabled behavior (equivalent to `ENABLE`), and `TARGET_LAG` is required.

Reviewer: ```suggestion
:::info dbt default differs from Snowflake's native default
In Snowflake's native DDL, omitting `SCHEDULER` defaults to `ENABLE`, and `TARGET_LAG` is required.
```

> In dbt, the default is **`DISABLE`**. If you do not specify `scheduler` or `target_lag`, dbt creates the dynamic table with `SCHEDULER = DISABLE` and takes responsibility for triggering refreshes during model execution.

Reviewer: ```suggestion
In dbt, the default value is `DISABLE`. If neither `scheduler` nor `target_lag` is specified, dbt creates the dynamic table with `scheduler: DISABLE` and manages refreshes directly.
```

> If you specify `target_lag` without explicitly setting `scheduler`, dbt sets `scheduler = ENABLE`.

Reviewer: ```suggestion
If you specify `target_lag` without explicitly setting `scheduler`, dbt sets `scheduler: ENABLE`.
```

:::

**Key points:**

> - `target_lag` can only be used when `scheduler` is set to `ENABLE`. Setting `target_lag` when `scheduler='DISABLE'` results in an error.

Reviewer: ```suggestion
- You can only use `target_lag` when `scheduler` is set to `ENABLE`. Setting `target_lag` when `scheduler: DISABLE'` results in an error.
```

> - When `scheduler='DISABLE'`, a manual refresh does **not** automatically refresh upstream dynamic table dependencies. This creates an isolation boundary, allowing dbt to manage specific table refreshes without triggering the entire pipeline. This is unlike `ENABLE` mode, where Snowflake cascades refreshes across the dependency graph.

Reviewer: ```suggestion
- When `scheduler: DISABLE`, a manual refresh does _not_ automatically refresh upstream dynamic table dependencies. This creates an isolation boundary, allowing dbt to manage specific table refreshes without triggering the entire pipeline. In contrast, `ENABLE` cascades refreshes across the dependency graph.
```

> - If a dynamic table with `scheduler='DISABLE'` depends on other dynamic tables, those upstream tables will not be refreshed when the downstream table is refreshed. The orchestrator (dbt) must manage the refresh order explicitly.

Reviewer: ```suggestion
- If a dynamic table with `scheduler: DISABLE` depends on other dynamic tables, those upstream tables will not be refreshed when the downstream table is refreshed. dbt must manage the refresh order explicitly.
```

For example, to let dbt manage refreshes (default behavior):

```sql
{{ config(
    materialized='dynamic_table',
    snowflake_warehouse='MY_WH',
) }}

select * from {{ source('raw', 'events') }}
```

To enable Snowflake-managed scheduling with a target lag:

```sql
{{ config(
    materialized='dynamic_table',
    snowflake_warehouse='MY_WH',
    scheduler='ENABLE',
    target_lag='5 minutes',
) }}

select * from {{ source('raw', 'events') }}
```

> Learn more about `SCHEDULER` in [Snowflake's docs](https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table#optional-parameters).

Reviewer: ```suggestion
Learn more about `scheduler` in [Snowflake's docs](https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table#optional-parameters).
```

<VersionBlock firstVersion="1.12">

#### How `target_lag` interacts with `scheduler`

`target_lag` works with [`scheduler`](#scheduler) to determine how dynamic table refreshes are managed:

<SimpleTable>

| `target_lag` | `scheduler` | Behavior |
|---|---|---|
| Set | `ENABLE` or omitted | Snowflake manages refreshes automatically. If `scheduler` is omitted, dbt defaults to `ENABLE`. |
| Not set | `DISABLE` or omitted | dbt manages refreshes during model execution. If `scheduler` is omitted, dbt defaults to `DISABLE`. |
| Set | `DISABLE` | Invalid: `DISABLE` does not accept `target_lag`. dbt raises an error. |
| Not set | `ENABLE` | Invalid: `ENABLE` requires `target_lag`. dbt raises an error. |

> </SimpleTable>

Author: crisp and clear thank you
